src/Prometheus/Storage/RedisTxn.php

```diff
@@ -0,0 +1,742 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Prometheus\Storage;
+
+use InvalidArgumentException;
+use Prometheus\Counter;
+use Prometheus\Exception\StorageException;
+use Prometheus\Gauge;
+use Prometheus\Histogram;
+use Prometheus\MetricFamilySamples;
+use Prometheus\Storage\RedisTxn\Metadata;
+use Prometheus\Storage\RedisTxn\Metric;
+use Prometheus\Summary;
+use RedisException;
+use RuntimeException;
+use function \sort;
+
+/**
+ * This is a storage adapter that persists Prometheus metrics in Redis.
+ *
+ * This library currently has two alternative Redis adapters:
+ * - {@see \Prometheus\Storage\Redis}: Initial Redis adapter written for this library.
+ * - {@see \Prometheus\Storage\RedisNg}: "Next-generation" adapter refactored to avoid use of the KEYS command to improve performance.
+ *
+ * While the next-generation adapter was an enormous performance improvement over the first, it still suffers from
+ * performance degradation that scales significantly as the number of metrics grows. This is largely due to the fact
+ * that the "collect" phase for metrics generally involves at least one network request per metric of each type.
+ *
+ * This adapter refactors the {@see \Prometheus\Storage\RedisNg} adapter to generally try and execute the "update" and
+ * "collect" operations of each metric type within a single Redis transaction.
+ *
+ * @todo Only summary metrics have been refactored so far. Complete refactor for counter, gauge, and histogram metrics.
+ */
+class RedisTxn implements Adapter
```

Comment:
This adapter is based on `RedisNg` and currently, the only changes are in the following methods:
- `updateSummary()`
- `collectSummaries()`

If there is interest from the repo maintainers, I'd like to apply the same style of refactoring for the remaining metric types:
- counters
- gauges
- histograms

Reply:
All metric types have been optimized. Based on benchmark timing, there may be further opportunity to optimize histograms.